conduit/transforms/audio.py

```diff
@@ -0,0 +1,352 @@
+        # band_edges_mel[i+1].  Thus, we need num_mel_bins + 2 values in
+        # the band_edges_mel arrays.
+        band_edges_mel = np.linspace(
+            self.hertz_to_mel(lower_edge_hertz),
+            self.hertz_to_mel(upper_edge_hertz),
+            num_mel_bins + 2,
+        )
+        # Matrix to post-multiply feature arrays whose rows are num_spectrogram_bins
+        # of spectrogram values.
+        mel_weights_matrix = np.empty((num_spectrogram_bins, num_mel_bins))
+        for i in range(num_mel_bins):
+            lower_edge_mel, center_mel, upper_edge_mel = band_edges_mel[i : i + 3]
+            # Calculate lower and upper slopes for every spectrogram bin.
+            # Line segments are linear in the *mel* domain, not hertz.
+            lower_slope = (spectrogram_bins_mel - lower_edge_mel) / (center_mel - lower_edge_mel)
+            upper_slope = (upper_edge_mel - spectrogram_bins_mel) / (upper_edge_mel - center_mel)
+            # .. then intersect them with each other and zero.
+            mel_weights_matrix[:, i] = np.maximum(0.0, np.minimum(lower_slope, upper_slope))
+        # HTK excludes the spectrogram DC bin; make sure it always gets a zero
+        # coefficient.
+        mel_weights_matrix[0, :] = 0.0
+        return mel_weights_matrix
+
+    def hertz_to_mel(self, frequencies_hertz: Union[float, np.ndarray]) -> np.ndarray:
+        """Convert frequencies to mel scale using HTK formula.
+        Args:
+          frequencies_hertz: Scalar or np.array of frequencies in hertz.
+        Returns:
+          Object of same size as frequencies_hertz containing corresponding values
+          on the mel scale.
+        """
+        return self.mel_high_freq_q * np.log(1.0 + (frequencies_hertz / self.mel_break_freq_hertz))
+
+
+class Framing(nn.Module):
+    def __init__(
+        self,
+        stft_hop_length_seconds: float = 0.010,
+        example_window_seconds: float = 0.96,  # Each example contains 96 10ms frames
+        example_hop_seconds: float = 0.96,  # with zero overlap.
+    ):
+        super().__init__()
+        self.stft_hop_length_seconds = stft_hop_length_seconds
+        self.example_window_seconds = example_window_seconds
+        self.example_hop_seconds = example_hop_seconds
+
+    def forward(self, data: torch.Tensor) -> torch.Tensor:
+        # Frame features into examples.
+        features_sample_rate = 1.0 / self.stft_hop_length_seconds
+        window_length = int(round(self.example_window_seconds * features_sample_rate))
+        hop_length = int(round(self.example_hop_seconds * features_sample_rate))
+
+        _, _, num_samples = data.shape
+        num_frames = 1 + int(np.floor((num_samples - window_length) / hop_length))
+        strides = (1, 1, hop_length)
+
+        return (
+            data.unsqueeze(1)
+            .as_strided(size=(num_frames, 64, window_length), stride=strides)
+            .unsqueeze(1)
+            .permute(0, 1, 3, 2)
+        )
```

Comment:
I am not at all confident about this

Reply:
You can just use ```movedim`` here as you're only permuting two dimensions

Reply:
That aside, I have no idea as to the correctness of this snippet